### PR TITLE
Narrow by clause expressions in switches with `true` condition

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/37178

### CODE REVIEW

**Comment by fatcerberus** [comment]
I personally hate this pattern (it feels like the bastard child of JS `switch..case` and Haskell `case` expressions) but people ask for TS to support it often enough that I'm not going to argue 😝

I guess it could save you from typing a bunch of parentheses... ~~in exchange for a bunch of `break` statements~~ 🚎

**Comment by codingedgar** [comment]
> I personally hate this pattern (it feels like the bastard child of JS `switch..case` and Haskell `case` expressions) but people ask for TS to support it often enough that I'm not going to argue 😝
> 
> 
> 
> I guess it could save you from typing a bunch of parentheses... ~~in exchange for a bunch of `break` statements~~ 🚎

Seems to be valid js tho https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#an_alternative_to_if...else_chains

supporting is nice 👌

**Comment by fatcerberus** [comment]
To be fair, [the bar is usually a *wee* bit higher than "it works in JS"](https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean) :wink: but yes, I agree that if people are regularly writing code like this, it's nice to support it.

**Comment by jakebailey** [comment]
Is the issue actually accepted?

**Comment by jakebailey** [comment]
(Code looks good, though.)

**Comment by Andarist** [comment]
Technically it doesnt have the corresponding label - if that’s what you are asking about. This feature was requested multiple times throughout the years and its a pattern mentioned in MDN docs so i jumped the gun on it and just implemented it as an excercise to learn more about CFA implementation in the compiler

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -26682,6 +26682,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             else if (expr.kind === SyntaxKind.TypeOfExpression && isMatchingReference(reference, (expr as TypeOfExpression).expression)) {
                 type = narrowTypeBySwitchOnTypeOf(type, flow.switchStatement, flow.clauseStart, flow.clauseEnd);
             }
+            else if (expr.kind === SyntaxKind.TrueKeyword) {
+                const clause = flow.switchStatement.caseBlock.clauses.find((_, index) => index === flow.clauseStart);
+                const clauseExpression = clause && clause.kind === SyntaxKind.CaseClause ? clause.expression : undefined;
+                if (clauseExpression) {
+                    type = narrowType(type, clauseExpression, /*assumeTrue*/ true);
```
Comment (Andarist) [inline src/compiler/checker.ts:26689]: I should probably use `narrowTypeByEquality` here or something and add corresponding tests

**Comment by jakebailey** [comment]
@typescript-bot perf test this

File: src/compiler/checker.ts (hunk 1 above)
Comment (Andarist) [reply src/compiler/checker.ts]: I re-examined this and I don't think this is needed anyhow here.

**Comment by jakebailey** [comment]
@typescript-bot test top200
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot pack this

**Review by jakebailey — APPROVED** [review]
LGTM so long as everything comes back clean.

**Comment by jakebailey** [comment]
@typescript-bot run dt

**Comment by jakebailey** [comment]
@typescript-bot run dt

**Comment by joscha** [comment]
Well done @Andarist, thank you!

**Pull request merged**